bftengine/src/bftengine/ClientsManager.hpp

```diff
@@ -218,9 +219,45 @@ class ClientsManager : public ResPagesClient<ClientsManager>, public IPendingReq
     bool committed = false;
   };
 
+  class RequestsInfo {
+   public:
+    void emplaceSafe(NodeIdType clientId, ReqId reqSeqNum, const std::string& cid);
+    bool removeRequestsOutOfBatchBoundsSafe(NodeIdType clientId, ReqId reqSequenceNum);
+    bool findSafe(ReqId reqSeqNum);
+    void clearSafe();
+    void removeOldPendingReqsSafe(NodeIdType clientId, ReqId reqSeqNum);
+    void removePendingForExecutionRequestSafe(NodeIdType clientId, ReqId reqSeqNum);
+
+    size_t size() const { return requestsMap_.size(); }
+    bool find(ReqId reqSeqNum) const;
+    bool isPending(ReqId reqSeqNum) const;
+    void markRequestAsCommitted(NodeIdType clientId, ReqId reqSeqNum);
+    void infoOfEarliestPendingRequest(Time& earliestTime, RequestInfo& earliestPendingReqInfo) const;
+    void logAllPendingRequestsExceedingThreshold(const int64_t threshold,
+                                                 const Time& currTime,
+                                                 int& numExceeding) const;
+
+   public:
+    std::mutex requestsMapMutex_;
+    std::map<ReqId, RequestInfo> requestsMap_;
+  };
+
+  class RepliesInfo {
+   public:
+    void deleteOldestReplyIfNeededSafe(NodeIdType clientId, uint16_t maxNumOfReqsPerClient);
+    bool insertOrAssignSafe(ReqId reqSeqNum, Time time);
+    bool findSafe(ReqId reqSeqNum);
+
+    bool find(ReqId reqSeqNum) const;
+
+   public:
+    std::mutex repliesMapMutex_;
```

Comment:
nit: It's better to declare mutex as mutable so that functions that should be const remain const. I think the hasReply is non-const now due to this mutex.
mutable std::mutex repliesMapMutex_;

Reply:
Good point.